cmake/CeleritasUtils.cmake

```diff
@@ -221,3 +254,66 @@ function(celeritas_configure_file input output)
 endfunction()
 
 #-----------------------------------------------------------------------------#
+
+macro(celeritas_setup_option var option) #[condition]
+  if(${ARGC} EQUAL 2)
+    # always-on-option
+    list(APPEND ${var}_OPTIONS ${option})
+  elseif(${ARGV2})
+    # variable evaluates to true
+    list(APPEND ${var}_OPTIONS ${option})
+  else()
+    list(APPEND ${var}_DISABLED_OPTIONS ${option})
+  endif()
+endmacro()
+
+#-----------------------------------------------------------------------------#
+
+function(celeritas_define_options var doc)
+  if(NOT ${var}_OPTIONS)
+    message(FATAL_ERROR "${var}_OPTIONS has no options")
+  endif()
+  mark_as_advanced(${var}_OPTIONS)
+
+  list(GET ${var}_OPTIONS 0 _default)
+  set(${var} "${_default}" CACHE STRING "${doc}")
+  set_property(CACHE ${var} PROPERTY STRINGS "${${var}_OPTIONS}")
+
+  list(FIND ${var}_OPTIONS "${${var}}" _index)
+  if(_index EQUAL -1)
+    message(SEND_ERROR "Invalid value ${var}=${${var}}: must be "
+      "${${var}_OPTIONS}; overriding for next configure")
```

Comment:
Consider making this fatal rather than 'fixing' it.  If for some reason the message is 'lost' and/or missed by the user (eg. terminal windows dies after cmake finished but before the user reads it or lost in a flood or errors), they will not notice that the build is working but with an option that is (arbitrarily) different from their selection.

Reply:
Yeah, I think this happened to Guilherme... I like the idea that successive runs of cmake will resolve conflicting configure options, but I see what you mean. I'll think about it.

Reply:
OK @pcanal , now this prints an error message:
```
CMake Error at cmake/CeleritasUtils.cmake:268 (message):
  Invalid selection not in list: "ORANGE" (CELERITAS_GEO=VecGeom): run `cmake
  -DCELERITAS_GEO=ORANGE` for a possible fix
```

We could make it smarter by querying the cache variable properties to maybe deduce if the variable was set by the user (don't override) versus an internal default (do override) but that can be deferred.

Reply:
It should be there: https://github.com/celeritas-project/celeritas/pull/726/commits/a80a5f7e35ecb39b5835bd593aa058827e6e06f7: the quoted is the list (which in this case has only one entry, so not sure what to do with that). The actual value is in parens.

Reply:
ok. I see.  Maybe consider this variation:
```
The selection "CELERITAS_GEO=VecGeom" is invalid; 
   the value must be in the following list: "ORANGE";
   run `cmake -DCELERITAS_GEO=ORANGE` for a possible fix
```

Reply:
Unfortunately I can't do that while reusing the macro across all the "invalid option" cases... how about:
```
CMake Error at cmake/CeleritasUtils.cmake:268 (message):
  Invalid option for CELERITAS_CORE_GEO: must be in the list [ORANGE]
  (CELERITAS_CORE_GEO=VecGeom): run `cmake -DCELERITAS_CORE_GEO=ORANGE` for a
  possible fix
```

Reply:
> I can't do that while reusing the macro across all the "invalid option" cases

I don't quite understand. I thought my proposal was only a reordering of the words and variables.  What am I missing?

Reply:
The macro `celeritas_error_incompatible_option`, which abstracts the decision to change parameter vs warning , displays the variable name and its value after a prespecified message:
```cmake
function(celeritas_error_incompatible_option msg var new_value)
  message(SEND_ERROR "${msg} (${var}=${${var}}):"
    "  run `cmake -D${var}=${new_value}` for a possible fix"
  )
endfunction()
```
remember that this macro is also used in a couple other places:
```cmake
  if(NOT Doxygen_FOUND)
    celeritas_error_incompatible_option(
      "Doxygen is required for building documentation but was not found"
      "CELERITAS_BUILD_DOCS" OFF
    )
  endif()
```
and
```cmake
    celeritas_error_incompatible_option(
      "CUDA mismatch between the VecGeom installation "
      "at ${VECGEOM_INSTALL_DIR} and Celeritas"
      CELERITAS_USE_CUDA
      "${VecGeom_CUDA_FOUND}"
    )
```

Reply:
The wording @sethrj propose at https://github.com/celeritas-project/celeritas/pull/726#issuecomment-1523470906 in conjunction with the update https://github.com/celeritas-project/celeritas/pull/726/files#r1177952200 would means a text like:
```
CMake Error at cmake/CeleritasUtils.cmake:268 (message):
  Invalid option for CELERITAS_CORE_GEO: must be in the list [ORANGE]
  (This is caused by CELERITAS_CORE_GEO being set to VecGeom): 
  run `cmake -DCELERITAS_CORE_GEO=ORANGE` for a possible fix
```
looks clear enough to me.

Reply:
< Invalid option for CELERITAS_CORE_GEO: must be in the list [ORANGE]
> Invalid option for CELERITAS_CORE_GEO, it must be in the list [ORANGE]

Reply:
Your message has a comma splice 😉 and I prefer error messages not to be so prosaic. I think this should satisfy us both:
```
CMake Error at cmake/CeleritasUtils.cmake:268 (message):
  Invalid setting CELERITAS_CORE_GEO=VecGeom: valid options are {ORANGE}

      Possible fix: cmake -DCELERITAS_CORE_GEO=ORANGE /Users/seth/.local/src/celeritas-temp/build
Call Stack (most recent call first):
  cmake/CeleritasUtils.cmake:302 (celeritas_error_incompatible_option)
  CMakeLists.txt:120 (celeritas_define_options)
```

Reply:
> a comma splice 

A period would do just fine then :)

Reply:
>  I think this should satisfy us both:

Yes, I agree :)